### PR TITLE
GH-73435: Add `pathlib.PurePath.full_match()`

### PR DESCRIPTION
**Pull request by barneygale**
In #101398 we added support for the recursive wildcard `**` in `pathlib.PurePath.match()`. This should allow arbitrary prefix and suffix matching, like `p.match('foo/**')` or `p.match('**/foo')`, but there's a problem: for relative patterns only, `match()` implicitly inserts a `**` token on the left hand side, causing all patterns to match from the right. As a result, it's impossible to match relative patterns from the left: `PurePath('foo/bar').match('bar/**')` is true!

This PR reverts the changes to `match()`, and instead adds a new `full_match()` method that:

- Allows empty patterns
- Supports the recursive wildcard `**`
- Matches the *entire* path when given a relative pattern

---

* Issue: gh-73435

----
📚 Documentation preview 📚: https://cpython-previews--114350.org.readthedocs.build/en/114350/library/pathlib.html#pathlib.PurePath.globmatch

### CODE REVIEW

File: Doc/library/pathlib.rst [hunk 1]
```diff
@@ -552,55 +552,54 @@ Pure paths provide the following methods and properties:
       PureWindowsPath('c:/Program Files')
 
 
-.. method:: PurePath.match(pattern, *, case_sensitive=None)
+.. method:: PurePath.globmatch(pattern, *, case_sensitive=None)
```
Comment (hugovk) [inline Doc/library/pathlib.rst]: Use an underscore for readability and accessibility?

```suggestion
.. method:: PurePath.glob_match(pattern, *, case_sensitive=None)
```

File: Lib/pathlib/_abc.py [hunk 2]
```diff
@@ -450,13 +450,29 @@ def match(self, path_pattern, *, case_sensitive=None):
         if case_sensitive is None:
             case_sensitive = _is_case_sensitive(self.pathmod)
         sep = path_pattern.pathmod.sep
-        if path_pattern.anchor:
-            pattern_str = str(path_pattern)
-        elif path_pattern.parts:
-            pattern_str = str('**' / path_pattern)
-        else:
+        our_parts = self.parts[::-1]
+        pat_parts = path_pattern.parts[::-1]
```
Comment (hugovk) [inline Lib/pathlib/_abc.py]: Use full word so no need to guess about later uses?
```suggestion
        pattern_parts = path_pattern.parts[::-1]
```

File: Doc/library/pathlib.rst (hunk 1 above)
Comment (barneygale) [reply Doc/library/pathlib.rst]: A couple other options: 

- `match_glob()` has a slightly more natural word order IMO
- `full_match()` aligns with `re.fullmatch()`.

Comment (hugovk) [reply Doc/library/pathlib.rst]: Yes, they sound good too.

**Comment by zooba** [comment]
Does the problem extend beyond "`**` is allowed to match nothing"? Or can we change the existing algorithm to require that a trailing `**` has to match at least one segment?

Presumably `.match("foo\\*")` has to find something after `foo\`, so would be a better version of `"foo\\**"` anyway, but I can't imagine anyone being happy that `"foo\\**"` matches `"foo\\"` - you wouldn't specify the `**` if you didn't want _something_ to be found there.

**Comment by barneygale** [comment]
`**` matching nothing is quite normal I think - it's what `glob.glob()`, `pathlib.Path.glob()`, zsh, and IIRC bash do. If you want at least one segment after `foo/`, you can spell that with `foo/**/*`.

**Comment by barneygale** [comment]
e.g. [zsh man page](https://linux.die.net/man/1/zshexpn0):

> A pathname component of the form '(foo/)#' matches a path consisting of zero or more directories matching the pattern foo.
> 
> As a shorthand, '**/' is equivalent to '(*/)#'; note that this therefore matches files in the current directory as well as subdirectories. Thus:
> 
>     ls (*/)#bar
> 
> or
> 
>     ls **/bar
> 
> does a recursive directory search for files named 'bar' (potentially including the file 'bar' in the current directory).

**Comment by zooba** [comment]
Okay, I definitely like having `full` in the name somewhere then. It's still a match, and the parallel is fullmatch is way more helpful than to glob (which would just raise questions about "how does `match` do its matching then?")

File: Lib/pathlib/_abc.py [hunk 3]
```diff
@@ -48,15 +48,15 @@ def _is_case_sensitive(pathmod):
 
 
 @functools.lru_cache(maxsize=256)
```
Comment (zooba) [inline Lib/pathlib/_abc.py]: Should we increase this, now that there are twice as many potential arguments?

**Review by zooba — COMMENTED** [review]
LGTM

**Comment by barneygale** [comment]
Thanks for the reviews, both.

**Pull request merged**